### ISSUE TITLE
Pass custom properties to shortcut constructors, reduce arity

### ISSUE THREAD

**Issue by jonchurch** [jshttp/http-errors#104]
Based on #103 

Users likely expect to be able to pass custom properties to errors when creating them via the shortcut methods such as `createError.NotFound()`.

Proposed change would look something like:

```js
const createError = require('http-errors')

createError.NotFound({message: "Aw shucks", data: { foo: "bar" } })
```

Today that does not work. You need to use the `createError({...})` form to pass custom properties to an error.

```js
createError(404, {message: "Aw shucks", data: { foo: "bar" } })
```

There's a few caveats with the current implementation, namely you can't set `status` via properties. 

```js
createError({ status: 404, message: "Aw shucks", data: { foo: "bar" } })
// status is ignored, will create a 500 server error with provided message and data fields
```

There is also a very high arity for `createError`, which should be locked down into a stable interface.

([code](https://github.com/jshttp/http-errors/blob/206aa2c15635dc1212c06c279540972aa90e23ea/index.js#L55C3-L70C4))
```js
for (var i = 0; i < arguments.length; i++) {
    var arg = arguments[i]
    var type = typeof arg
    if (type === 'object' && arg instanceof Error) {
      err = arg
      status = err.status || err.statusCode || status
    } else if (type === 'number' && i === 0) {
      status = arg
    } else if (type === 'string') {
      msg = arg
    } else if (type === 'object') {
      props = arg
    } else {
      throw new TypeError('argument #' + (i + 1) + ' unsupported type ' + type)
    }
  }
```

Any of these changes would be semver major. There aren't plans for a major release of `http-errors` yet, but these are things Im thinking about.

**Comment by jonchurch** [comment]
There's been prior discussions around the semantics of the inputs we accept, and how it impacts the outputs:

#58 #37 and likely more

**Comment by jonchurch** [comment]
Yes, Im saying that any of these changes would indeed be semver major.

Im in no rush to make this change or the subclassing change.

This is a tracking issue for a change
> related to the interfaces exposed by the library for accepting arguments, which can use some rethinking.

if we did want to land a change in this category, it would need to be in a major.

**Comment by jonchurch** [comment]
Related #96 #78

**Comment by mahmoodhamdi** [comment]
The current loose arity is convenient but definitely surprising in edge cases. I've seen this bite people when passing an error object that already has a status — the error's status silently wins over the explicit status argument (which is what #96 describes).

For v3, I think the named constructor approach would benefit from an options object pattern (as @jonchurch described), similar to how the native `Error` constructor accepts `{ cause }` in newer Node.js:

```js
// Current — message only
new createError.NotFound('Resource not found')

// Proposed — options object as second arg, aligned with Error(msg, { cause })
new createError.NotFound('Resource not found', {
  expose: true,
  headers: { 'retry-after': '30' },
  cause: originalError
})
```

This would:
1. Solve #78 (expose on named constructors)
2. Align with the native `Error(message, options)` pattern from ES2022
3. Keep backwards compatibility for the `(message)` signature
4. Avoid the ambiguity issues with the current `createError()` factory

For the factory function itself, locking down to explicit signatures instead of the current argument-type-sniffing would definitely reduce surprises. Something like:

```js
createError(status)
createError(status, message)
createError(status, message, properties)
createError(status, error)
createError(status, error, properties)
```

Where `status` is always required as the first argument — no more defaulting to 500 based on argument types.
